Create index.eng.md (#98)
planetarium/snack.planetarium.dev: content/posts/2020/04/rocksdb/index.eng.md

After reviewing various alternative storage methods, we decided that [RocksDB], a [Key-Value Database] library developed by Facebook, was our best option. We decided to create an `IStore` implementation called [`RocksDBStore`] to use as a backend. In this article, I would like to share our experiences in developing `RocksDBStore`.

[Libplanet]: https://libplanet.io/
[`IStore`]: https://docs.libplanet.io/0.8.0/api/Libplanet.Store.IStore.html
[`DefaultStore`]: https://docs.libplanet.io/0.8.0/api/Libplanet.Store.DefaultStore.html
[RocksDB]: https://rocksdb.org/
[Nine Chronicles]: https://nine-chronicles.com/
[Key-Value Database]: https://en.wikipedia.org/wiki/Key-value_database
[`RocksDBStore`]: https://github.com/planetarium/libplanet/blob/master/Libplanet.RocksDBStore/RocksDBStore.cs

## Including Dependent Libraries[^1]

> RocksDB relies on other libraries for compression or memory allocation purposes. Unlike [Windows build], for macOS and Linux, RocksDB native libraries must also be installed in the system to use them in the form of dynamic link libraries (*.so* and *.diib*).

Reviewer: RocksDB를 사용하기 위해서 의존 라이브러리들이 설치되어 있어야 한다는 의미로 쓴거였는데, 이런식으로 쓰면 맞으려나요?

```suggestion
RocksDB relies on other libraries for compression or memory allocation purposes. Unlike [Windows build], for macOS and Linux, the dependent libraries must also be installed in the system to use the RocksDB native library in the form of a dynamic link library (*.so* and .*dylib*).
```

Author: 네 이렇게 쓰면 말씀하신 의도대로 전달될 것 같습니다!

> RocksDB relies on other libraries for compression or memory allocation purposes. Unlike [Windows build], for macOS and Linux, the dependent libraries must also be installed in the system to use the RocksDB native library in the form of a dynamic link library (*.so* and .*dylib*).

Reviewer 2: ```suggestion
RocksDB relies on other libraries for compression or memory allocation purposes. Unlike [Windows build], for macOS and Linux, the dependent libraries must also be installed in the system to use the RocksDB native library in the form of dynamic link libraries (*.so* and .*dylib*).
```

In a typical server app, it is natural to install all dependent libraries in your system. This is because the system that runs the server app is usually operated only for that server app. But because we're building an app that serves as blockchain nodes and runs on gamers' systems, it's hard to ask all gamers to install these libraries separately.

So what we came up with was to put the libraries that RocksDB relies on in the game client and distribute them. However, building RocksDB in the form of a dynamic link library without any modifications caused the built RocksDB library to not be able to find the dependent libraries that were included in the game clients.

> To resolve this issue, we used a method of modifying [rpath] in the RocksDB dynamic link library file. rpath refers to <q>run-time search path</q>, which is hard-coded within a library file or exe file so the [dynamic linking] [loader] can find the required library in that file. Initially, we considered modifying the rpath when building the RocksDB library, but we eventually decided to modify the rpath in the completed library file because the build script in RocksDB turned out to be more complicated than we thought. Fortunately, with tools called [`install_name_tool`] on macOS and [`patchelf`] on Linux, you can simply modify the rpath to the directory where the current RocksDB library exists.

Reviewer: ```suggestion
To resolve this issue, we used a method of modifying [rpath] in the RocksDB dynamic link library file. rpath refers to <q>run-time search path</q>, which is hard-coded within a library file or executable file so the [dynamic linking] [loader] can find the required library in that file. Initially, we considered modifying the rpath when building the RocksDB library, but we eventually decided to modify the rpath in the completed library file because the build script in RocksDB turned out to be more complicated than we thought. Fortunately, with tools called [`install_name_tool`] on macOS and [`patchelf`] on Linux, you can simply modify the rpath to the directory where the current RocksDB library exists.
```

> ```
> # macOS
> $ install_name_tool -add-rpath '@loader_path' librocksdb.dylib
> 
> # linux
> $ patchelf --set-rpath $ORIGIN librocksdb.so
> ```

Reviewer 2: ```suggestion
~~~~ bash
# macOS
$ install_name_tool -add-rpath '@loader_path' librocksdb.dylib

# linux
$ patchelf --set-rpath $ORIGIN librocksdb.so
~~~~
```

For instance, since there is no method to count the number of rows of stored data, it is necessary to implement them using various ways, such as storing the number separately every time data is updated or counting while traversing the entire data on each update.

Another example is the key search feature. RocksDB's `Seek` takes the prefix of the key as input to locate the key. While it's easy to assume that this feature will only find keys that match the prefix like a typical database search, it's actually more similar to [`lseek()`][lseek(2)], which moves the offset of the file. Therefore, when using this feature to traverse a key, you need to check at each key that the first head of that key matches the string parts you are looking for.

[relational database]: https://en.wikipedia.org/wiki/Relational_database
[LiteDB]: https://www.litedb.org/
[lseek(2)]: http://man7.org/linux/man-pages/man2/lseek.2.html

## Common Mistakes when Overlooking the Docs

The APIs and documentation of RocksDB were not as user-friendly as expected, so extra attention was needed to use it.

> One example was the [Column Family], which acts like a namespace. After creating a column family in the database, we expected the column family to be brought up automatically when using the database again. However, there was an exception if we did not specify all column families in the database using the API called `ListColumnFamilies` when opening the database.

Reviewer: ```suggestion
One example was the [Column Family], which acts like a namespace. After creating a column family in the database, we expected the column family to be brought up automatically when using the database again. However, an exception occurs if we did not specify all column families in the database using the API called `ListColumnFamilies` when opening the database.
```

---
title: Applying RocksDB to Libplanet
date: 2020-04-17
authors: [seunghun.lee]
translators: [kidon.seo]
---

Hello, I'm Seunghun Lee, [Libplanet] developer at Planetarium.

> Libplanet provides a storage layer abstraction interface called [`IStore`]and its basic implementation called [`DefaultStore`]. `DefaultStore` had been used to develop [Nine Chronicles] and although it was included as a base in Libplanet and had the upside of being able to use it immediately, there were certainly limitations in terms of performance and storage efficiency.

Reviewer 3: ```suggestion
Libplanet provides a storage layer abstraction interface called [`IStore`] and its basic implementation called [`DefaultStore`]. `DefaultStore` had been used to develop [Nine Chronicles] and although it was included as a base in Libplanet and had the upside of being able to use it immediately, there were certainly limitations in terms of performance and storage efficiency.
```
